simple/app.go

```diff
@@ -0,0 +1,334 @@
+package simple
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+
+	"github.com/grafana/grafana-app-sdk/app"
+	"github.com/grafana/grafana-app-sdk/k8s"
+	"github.com/grafana/grafana-app-sdk/operator"
+	"github.com/grafana/grafana-app-sdk/resource"
+)
+
+var _ app.Provider = &AppProvider{}
+
+type Converter k8s.Converter
+
+// AppProvider is a simple implementation of app.Provider which returns AppManifest when Manifest is called,
+// and calls NewAppFunc when NewApp is called.
+type AppProvider struct {
+	AppManifest app.Manifest
+	NewAppFunc  func(config app.Config) (app.App, error)
+}
+
+// Manifest returns the AppManifest in the AppProvider
+func (a *AppProvider) Manifest() app.Manifest {
+	return a.AppManifest
+}
+
+// NewApp calls NewAppFunc and returns the result
+func (a *AppProvider) NewApp(settings app.Config) (app.App, error) {
+	return a.NewAppFunc(settings)
+}
+
+// NewAppProvider is a convenience method for creating a new AppProvider
+func NewAppProvider(manifest app.Manifest, newAppFunc func(cfg app.Config) (app.App, error)) *AppProvider {
+	return &AppProvider{
+		AppManifest: manifest,
+		NewAppFunc:  newAppFunc,
+	}
+}
+
+var (
+	_ app.App = &App{}
+)
+
+// App is a simple, opinionated implementation of resource.App.
+// It must be created with NewApp to be valid.
+type App struct {
+	informerController *operator.InformerController
+	operator           *operator.Operator
+	clientGenerator    resource.ClientGenerator
+	kinds              map[string]AppManagedKind
+	internalKinds      map[string]resource.Kind
+	cfg                AppConfig
+	converters         map[string]Converter
+}
+
+// AppConfig is the configuration used by App
+type AppConfig struct {
+	Name           string
+	KubeConfig     rest.Config
+	InformerConfig AppInformerConfig
+	ManagedKinds   []AppManagedKind
+	Converters     map[schema.GroupKind]Converter
+}
+
+// AppInformerConfig contains configuration for the App's internal operator.InformerController
+type AppInformerConfig struct {
+	ErrorHandler       func(context.Context, error)
+	RetryPolicy        operator.RetryPolicy
+	RetryDequeuePolicy operator.RetryDequeuePolicy
+	FinalizerSupplier  operator.FinalizerSupplier
+}
+
+// AppManagedKind is a Kind and associated functionality used by an App.
+type AppManagedKind struct {
+	// Kind is the resource.Kind being managed. This is equivalent to a kubernetes GroupVersionKind
+	Kind resource.Kind
+	// Reconciler is an optional reconciler to run for this Kind. Only one version of a Kind should have a Reconciler,
+	// otherwise, duplicate events will be received.
+	Reconciler operator.Reconciler
+	// Watcher is an optional Watcher to run for this Kind. Only one version of a Kind should have a Watcher,
+	// otherwise, duplicate events will be received.
+	Watcher operator.ResourceWatcher
```

Comment:
This should be either–or, right? Should we consider dropping `Watcher` from here, since we've discussed moving to `Reconciler` long-term?

Reply:
Yeah, this is an either-or (if you use both, you'll get duplicated events). I'm not sure about dropping Watcher yet as it seems like most new apps prefer the Watcher approach, I think the first step might be to have the generated code prefer reconcilers.

Reply:
Let's make sure then that we validate that only either one is passed.